### ISSUE TITLE
Deadwood isn't chasing CNAMEs Correctly?

### ISSUE THREAD

**Issue by kentborg** [samboy/MaraDNS#26]
Deadwood isn't giving me the same answer as does Google's 8.8.8.8.

From Deadwood: "dig @127.0.0.1 www.urban-radish.com"

I get:

;; ANSWER SECTION:
www.urban-radish.com.	3016	IN	CNAME	urban-radish.squarespace.com.
urban-radish.squarespace.com. 3016 IN	A	141.8.225.68

But from Google: "dig @8.8.8.8 www.urban-radish.com"

I get:

;; ANSWER SECTION:
www.urban-radish.com.	14005	IN	CNAME	urban-radish.squarespace.com.
urban-radish.squarespace.com. 299 IN	A	198.185.159.176
urban-radish.squarespace.com. 299 IN	A	198.49.23.177
urban-radish.squarespace.com. 299 IN	A	198.185.159.177

However, explicitly looking up the CNAME works:

"dig @127.0.0.1 urban-radish.squarespace.com"

I get:

;; ANSWER SECTION:
urban-radish.squarespace.com. 300 IN	A	198.49.23.177
urban-radish.squarespace.com. 300 IN	A	198.185.159.176
urban-radish.squarespace.com. 300 IN	A	198.185.159.177

Am I pointing to something  wrong? I noticed that h.root-servers.net seems wrong in DwSocket.h, per https://www.iana.org/domains/root/servers

Thanks,

-kb

**Comment by samboy** [comment]
I am not going to fix this issue.

Here’s what going on: The DNS specs do not clearly define how names should be resolved, and there are a lot of corner cases with multiple correct answers. This is one of those corner cases.

In more technical detail: There is an out-of-bailiwick A record attached to a CNAME record; since Deadwood has a secure way to resolve those kinds of records without having to chase down the CNAME record, Deadwood does that in order to solve a DNS record as quickly as possible. Other DNS servers choose to just discard the out of bailiwick record. Neither choice is the correct one, but with poorly maintained zones like urban-radish.com, different DNS servers will give different answers, since  216.21.231.99 has a different A record for urban-radish.squarespace.com than 208.78.70.6 does.

Here is the output of dig showing what’s going on:

    $ dig @216.21.231.99 www.urban-radish.com
    [...]
    ;; QUESTION SECTION:
    ;www.urban-radish.com.          IN      A

    ;; ANSWER SECTION:
    www.urban-radish.com.   14400   IN      CNAME   urban-radish.squarespace.com.
    urban-radish.squarespace.com. 3600 IN   A       141.8.225.68

The way to resolve this issue is to ask the domain administrator for urban-radish.com to fix or remove the urban-radish.squarespace.com A records on the servers for urban-radish.com.

As it turns out, handling this corner case by just discarding the out of bailiwick A record would probably make Deadwood’s code simpler and smaller, but the design for Deadwood’s resolver dates back to the original MaraDNS recursive resolver in 2001, when a lot of production DNS servers accepted that kind out of bailiwick A record, and changing Deadwood’s behavior may cause other issues which I do not have the resources to deal with. To be honest, at this point I only fix security issues or simply typo bugs in MaraDNS or Deadwood.

I have tried to find someone to step up to the plate as a maintainer for MaraDNS or to start a MaraDNS-ng fork; no one wants to do all that work for free. Welcome to the world of open source economics: The code is free, but there is no guarantee of support, since there is no money on the table compelling one to support the program.

Here is an excerpt from a [blog entry from last year](http://samiam.org/blog/2015-02-11.html):

The fact that BIND or whatever handles a particular RFC violating DNS packet a particular way does not compel me to have Deadwood handle that corner case the same way. Not unless there is significant money on the table.

Over the years, I have added countless fuzzy heuristics to Deadwood so that it can resolve as many names as possible. For years, I have said “If BIND or Unbound can resolve a given name, but Deadwood can not, that is a bug.”

But, my life has changed, and I need to put the foot down
